### PR TITLE
Fix crashes when trying to (re-)create shadows for destroyed windows

### PR DESCRIPTION
**Pull request by tryone144**
This should fix various crashes where the shadow images created/updated in-time (both legacy and experimental backends).

<details>
<summary>EDIT: Fix has already been merged outside of this PR</summary>

First of all, fixes https://github.com/yshui/picom/issues/465 by partially reverting https://github.com/yshui/picom/commit/32754b0 (this seems to be the "benefit to keep them separate").
Shadow images might be updated/invalidated (via the `SHADOW_STALE` flag) depending on the focus-state (in `recheck_focus()`). We therefore have to refresh/create shadow images after that. Handling of the `MAPPED` flag however still has to be done before focus checking, otherwise we introduce a single frame lag for windows that are to be mapped and have rules depending on the focus-state.
@yshui I hope the naming for the split flag_update function is okay.

</details>

Furthermore, by expanding https://github.com/yshui/picom/commit/f493447b33daf009cc0a15fa52f52115bbc91cc9 to clear the `SHADOW_STALE` flag as well we prevent recreation of shadow textures for windows that just got unmapped/destroyed.
~~@yshui Not sure if this is the right approach but this seems to do the trick (the comment suggest this is the reason `PIXMAP_STALE` is cleared as well).~~ Not quite the right way. Something like this seems to be part of the reason for #239.

At least https://github.com/yshui/picom/issues/394 seems to be caused by windows that get destroyed but for some reason their shadow images get invalidated and are to be recreated (~~which fails for a non-existent window~~). This sets the `SHADOW_NONE` directly but won't update the `SHADOW_NONE` and `SHADOW_STALE` flags since the window is already destroyed. I haven't dug deep enough yet but the windows causing it seem to trigger an event after they are unmapped/destroyed. Fade-out seems to exaggerate this problem. The offending windows then have both `SHADOW_NONE` and `SHADOW_STALE` flags set.

I'm going to ping possible related issues to confirm this fixed their crashes as well: https://github.com/yshui/picom/issues/390, https://github.com/yshui/picom/issues/393, https://github.com/yshui/picom/issues/396, https://github.com/yshui/picom/issues/425.

### CODE REVIEW

**Comment by tryone144** [comment]
Hold on, something new crept up with the tests for #239...

**Comment by tryone144** [comment]
Okay. Re-creating the shadow image does not seem to be the problem (we actually want to be able to create the shadow for unmapped windows).

Currently, the window receives a resize event and immediately gets destroyed. This causes the shadow image to be invalidated, but then the flags won't be updated on a destroyed window leading to an assertion failure even though the shadow image is present.

**Comment by yshui** [comment]
Thanks.

I hope one day we will be able to sort out the state changes and image updates. Right now everything is so messy and intertwined.

**Comment by yshui** [comment]
> we actually want to be able to create the shadow for unmapped windows

Can you explain why?

**Comment by tryone144** [comment]
> > we actually want to be able to create the shadow for unmapped windows
> 
> Can you explain why?

That's part of the fix for the mentioned issue: https://github.com/yshui/picom/commit/33a106e254c18d945a1c363a8b658f4c8d79c406
A window might be unmapped and fading out but we have to update the shadow due to some other rule (e.g. focus change?).

If you think the fix for https://github.com/yshui/picom/issues/465 is fine I can merge that one right now and keep this PR open for the other issues.

**Comment by yshui** [comment]
> That's part of the fix for the mentioned issue: 33a106e

Ah right. I forgot about that.

> If you think the fix for #465 is fine I can merge that one right now and keep this PR open for the other issues.

:+1:

**Comment by tryone144** [comment]
> I hope one day we will be able to sort out the state changes and image updates. Right now everything is so messy and intertwined.

Yeah, I think I found the reason for the other assertion failure: If a container window looses its child its state is re-evaluated. With [my config](https://pastebin.com/y2QsBa00) this means the container will loose its shadow (i.e. `SHADOW_STALE` flag gets set). Immediately after, the container window is destroyed as well with a shadow image in state `SHADOW_STALE`.
This seems to be another race-condition, as I can't trigger it with log-level `TRACE` and hardly with `DEBUG`.

**Comment by tryone144** [comment]
A related issue with the state updates yielding the same assertion failure: https://github.com/yshui/picom/issues/487#issuecomment-693351742

**Comment by frebib** [comment]
I have been running next `HEAD` with this applied for a few weeks now on two devices and crashes are significantly fewer than they were before. I still see rare crashes when changing dpi (reloading i3 and restarting a bunch of things at once- I think it's the same reason as #487). I'm experiencing some lock-ups on the laptop with Intel iGPU but I'm pretty sure that's related to the buggy gpu drivers more than anything.
Generally speaking though I think this is a big improvement

**Comment by tryone144** [comment]
Thanks for testing this @frebib. This currently seems to work around only one of many "messy and intertwined" path that cause this assertion to fail, in particular the one I frequently encountered with the mate-polkit-agent. There might be a cleaner fix for that, but this should do the trick for this particular issue. Still need to find a working solution for https://github.com/yshui/picom/issues/487.

**Comment by tryone144** [comment]
#394 (and likely similarly caused #396 and #425) should be fixed by not trying to recreate the shadow when a window is destroyed.
#487 (and the crash in #390 seems to be the same) should also be fixed by first updating the `STALE_CLIENT` and then updating the pixmap and shadow images as they might become invalidated by the updated client.

Thanks again @frebib and the others for testing these fixes. @yshui if you have no objection, I'll merge these into master.

**Comment by frebib** [comment]
I just found out that I am able to trigger the `WIN_FLAGS_PIXMAP_STALE` assertion with 24463c3 :(
Trace logs: https://gist.githubusercontent.com/frebib/935b287dc49cc28fe51f4e08ced48735/raw
It seems to only happen on my laptop, though.

**Comment by tryone144** [comment]
Do you have a backtrace for that crash? Might be the related to https://github.com/yshui/picom/issues/395.

**Comment by frebib** [comment]
Ah yeah no backtrace without the systemd-coredump output, apologies. Looks like it is the same assertion but a different codepath
```
Sep 28 09:15:20 frebib-Cf picom[873181]: [ 28/09/20 09:15:20.984 reset_enable INFO ] picom is resetting...
Sep 28 09:15:21 frebib-Cf picom[873181]: [ 28/09/20 09:15:21.072 gl_has_extension INFO ] Missing GL extension GL_GREMEDY_string_marker.
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.392 win_clear_flags WARN ] Flags cleared on a destroyed window 0x010000a4 (Alacritty)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.392 win_clear_flags WARN ] Flags cleared on a destroyed window 0x0100004a (Mozilla Firefox)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.392 win_clear_flags WARN ] Flags cleared on a destroyed window 0x01000058 (Alacritty)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.392 win_clear_flags WARN ] Flags cleared on a destroyed window 0x0100009c (polybar-i3bar_eDP1)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.522 win_clear_flags WARN ] Flags cleared on a destroyed window 0x010000a4 (Alacritty)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.522 win_clear_flags WARN ] Flags cleared on a destroyed window 0x0100004a (Mozilla Firefox)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.522 win_clear_flags WARN ] Flags cleared on a destroyed window 0x01000058 (Alacritty)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.522 win_clear_flags WARN ] Flags cleared on a destroyed window 0x0100009c (polybar-i3bar_eDP1)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.567 win_clear_flags WARN ] Flags cleared on a destroyed window 0x010000a4 (Alacritty)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.567 win_clear_flags WARN ] Flags cleared on a destroyed window 0x0100004a (Mozilla Firefox)
Sep 28 09:15:24 frebib-Cf picom[873181]: [ 28/09/20 09:15:24.567 win_clear_flags WARN ] Flags cleared on a destroyed window 0x0100009c (polybar-i3bar_eDP1)
Sep 28 09:15:25 frebib-Cf picom[873181]: [ 28/09/20 09:15:25.322 win_clear_flags WARN ] Flags cleared on a destroyed window 0x01000093 (polybar-i3bar_eDP1)
Sep 28 09:15:25 frebib-Cf picom[873181]: [ 28/09/20 09:15:25.851 win_clear_flags WARN ] Flags cleared on a destroyed window 0x01000093 (polybar-i3bar_eDP1)
Sep 28 09:15:27 frebib-Cf picom[873181]: [ 28/09/20 09:15:27.899 configure_root INFO ] Root configuration changed, new geometry: 4492x2160
Sep 28 09:15:27 frebib-Cf picom[873181]: picom: ../src/win.c:313: win_release_images: Assertion `!win_check_flags_all(w, WIN_FLAGS_PIXMAP_STALE)' failed.
Sep 28 09:15:28 frebib-Cf systemd[8843]: picom.service: Main process exited, code=dumped, status=6/ABRT
Sep 28 09:15:28 frebib-Cf systemd[8843]: picom.service: Failed with result 'core-dump'.
Sep 28 09:15:28 frebib-Cf systemd[8843]: picom.service: Scheduled restart job, restart counter is at 1.
Sep 28 09:15:28 frebib-Cf systemd[8843]: Stopped A lightweight compositor for X11.
```
```
Process 873181 (picom) of user 1000 dumped core.

Stack trace of thread 873181:
#0  0x00007fd9b590c615 raise (libc.so.6 + 0x3d615)
#1  0x00007fd9b58f5862 abort (libc.so.6 + 0x26862)
#2  0x00007fd9b58f5747 __assert_fail_base.cold (libc.so.6 + 0x26747)
#3  0x00007fd9b5904bf6 __assert_fail (libc.so.6 + 0x35bf6)
#4  0x0000563e6f30395e win_release_images (picom + 0x1095e)
#5  0x0000563e6f2fff59 destroy_backend (picom + 0xcf59)
#6  0x0000563e6f302a2b configure_root (picom + 0xfa2b)
#7  0x0000563e6f302dc9 draw_callback (picom + 0xfdc9)
#8  0x00007fd9b5e8f033 ev_invoke_pending (libev.so.4 + 0x5033)
#9  0x00007fd9b5e92902 ev_run (libev.so.4 + 0x8902)
#10 0x0000563e6f2fe707 session_run (picom + 0xb707)
#11 0x00007fd9b58f7152 __libc_start_main (libc.so.6 + 0x28152)
#12 0x0000563e6f2ffd5e _start (picom + 0xcd5e)
```

**Comment by tryone144** [comment]
Thanks. Your bt shows indeed a different codepath but being unable to release stale images when shutting down the backend seems to be the same problem, just a different reason we are shutting down.

In your case there is a race condition where the root window is reconfigured (most likely a size change) which will destroy and re-init the backend. If at the same time a window's size is changed its pixmap will be marked stale but updated only *after* the root reconfiguration has been handled. I can only trigger it in about one of 3 or 4 tries.

It should be safe to just clear the `STALE` flags when destroying the backend as we actually want to free all acquired images. If they are needed afterwards they will be reacquired by the backend initialization routine. If you want, you can try the following patch:
```diff
diff --git a/src/picom.c b/src/picom.c
index 9b7c534..9f870e9 100644
--- a/src/picom.c
+++ b/src/picom.c
@@ -401,6 +401,10 @@ static void destroy_backend(session_t *ps) {
 
 		if (ps->backend_data) {
 			if (w->state == WSTATE_MAPPED) {
+				// Ignore stale images as we want to free them anyway.
+				// They will be recreated in `initialize_backend()` later
+				// if needed.
+				win_clear_flags(w, WIN_FLAGS_IMAGES_STALE);
 				win_release_images(ps->backend_data, w);
 			} else {
 				assert(!w->win_image);
```

**Comment by frebib** [comment]
Applying that patch does stop the crashing on destroy, but I've managed to trigger yet another assertion in another place lol. I won't burden your PR here any longer. I'll open an issue for that one.

File: src/win.c [hunk 1]
```diff
@@ -1887,10 +1889,14 @@ bool destroy_win_start(session_t *ps, struct win *w) {
 	}
 
 	if (w->managed) {
-		// Clear PIXMAP_STALE flag, since the window is destroyed there is no
-		// pixmap available so STALE doesn't make sense.
+		// Clear IMAGES_STALE flags since the window is destroyed: Clear
+		// PIXMAP_STALE as there is no pixmap available anymore, so STALE doesn't
+		// make sense.
+		// XXX Clear SHADOW_STALE as setting/clearing flags on a destroyed window
```
Comment (yshui) [inline src/win.c:1895]: Maybe we could loose that check?

Comment (yshui) [reply src/win.c:1895]: Or maybe we should actually regenerate the shadow image here, as we clear the stale flag?

Comment (tryone144) [reply src/win.c:1895]: We are destroying the window at this point. We could "manually" recreate the shadow here if the flag is set. Updating the flags should work, as long as we haven't set `w->state = WSTATE_DESTROYING`.
Is there a specific reason the image updates are only done in the x critical section, that would be violated if we call the update here?

Comment (yshui) [reply src/win.c:1895]: I see, that make sense. Yeah, I think this is fine

**Comment by yshui** [comment]
@tryone144 Looks like after the new changes, we should clear some of the update flags when destroying a window.

**Comment by tryone144** [comment]
> @tryone144 Looks like after the new changes, we should clear some of the update flags when destroying a window.

As long as it doesn't cause too many visible artifacts with fading enabled, this should be fine. After all, the window is destroyed anyway.

@yshui What about https://github.com/yshui/picom/pull/479#issuecomment-700881298 related to #395? Can you take a look if that fix is (still) sufficient?

**Comment by yshui** [comment]
@tryone144 I will close #395. I will open a new issue if I see it again.

BTW, I did write up about the window state management here: https://github.com/yshui/picom/wiki/Window-states-management. Trying to clear my thoughts on how to make this less messy, but I don't know if the information is complete, or correct.

**Pull request merged**